test/unit/lib/classes/Variables.test.js

```diff
@@ -2780,4 +2708,113 @@ module.exports = {
       expect(processedConfig.custom.nestedReference).to.equal('resolvedNested');
     });
   });
+
+  describe('variable resolving', () => {
+    describe('when unresolvedVariablesNotificationMode is set to "error"', () => {
+      it('should error for missing "environment variable" type variables', async () => {
+        await expect(
+          runServerless({
+            fixture: 'variables',
+            cliArgs: ['print'],
+            configExt: {
+              unresolvedVariablesNotificationMode: 'error',
+              custom: { myVariable: '${env:missingEnvVar}' },
+            },
+          })
+        ).to.be.rejectedWith('environment variable');
+      });
+
+      it('should error for missing "option" type variables', async () => {
+        await expect(
+          runServerless({
+            fixture: 'variables',
+            cliArgs: ['print'],
+            configExt: {
+              unresolvedVariablesNotificationMode: 'error',
+              custom: { myVariable: '${opt:missingOpt}' },
+            },
+          })
+        ).to.be.rejectedWith('option');
+      });
+
+      it('should error for missing "service attribute" type variables', async () => {
+        await expect(
+          runServerless({
+            fixture: 'variables',
+            cliArgs: ['print'],
+            configExt: {
+              unresolvedVariablesNotificationMode: 'error',
+              custom: { myVariable: '${self:missingAttribute}' },
+            },
+          })
+        ).to.be.rejectedWith('service attribute');
+      });
+
+      it('should error for missing "file" type variables', async () => {
+        await expect(
+          runServerless({
+            fixture: 'variables',
+            cliArgs: ['print'],
+            configExt: {
+              unresolvedVariablesNotificationMode: 'error',
+              custom: { myVariable: '${file(./missingFile)}' },
+            },
+          })
+        ).to.be.rejectedWith('file');
+      });
+    });
+
+    describe('when unresolvedVariablesNotificationMode is set to "warn"', () => {
+      it('prints warnings to the console but no deprecation message', async () => {
+        const out = await runServerless({
+          fixture: 'variables',
+          cliArgs: ['print'],
+          configExt: {
+            unresolvedVariablesNotificationMode: 'warn',
+            custom: {
+              myVariable1: '${env:missingEnvVar}',
+              myVariable2: '${opt:missingOpt}',
+              myVariable3: '${self:missingAttribute}',
+              myVariable4: '${file(./missingFile)}',
+            },
+          },
+        });
+
+        ({ serverless } = out);
```

Comment:
Why not:

```javascript
const { serverless: { stdout } } = await runServerless(...);
```

Reply:
Because `serverless` is declared [here](https://github.com/serverless/serverless/blob/master/test/unit/lib/classes/Variables.test.js#L36) as a `let`, meaning this'll be a shadowing that eslint errors on.

Reply:
But in this case you don not access `serverless` but just `stdout` (and it's only that which I define in snippet)

Reply:
It's used on the line straight after this one - but I guess actually I could do:

```
const { serverless: { triggeredDeprecations }, stdoutData } = await runServerless(...);
```

Reply:
I've actually didn't notice need to access `serverless.triggeredDeprecations`, so sorry for misleading suggestion

Anyway it was flawed that `runServerless` tests share same context as other tests, it should not be the case. I've fixed that in `master`, and now there's no `serverless` shadowing, please rebase.

Reply:
Done :)